src/sync.ts

```diff
@@ -1604,7 +1604,13 @@ export class SyncApi {
                     success();
                 },
                 (err) => {
-                    if (err.httpStatus == 400 || err.httpStatus == 404) {
+                    if (!this.running) {
+                        clearTimeout(this.keepAliveTimer);
+                        if (this.connectionReturnedDefer) {
+                            this.connectionReturnedDefer.reject();
+                            this.connectionReturnedDefer = undefined;
+                        }
```

Comment:
>  In any case, the new code needs more comments to explain what is going on.

please do give this code some comments to explain what it is doing. Imagine you are unfamiliar with this code and trying to understand it. When does it happen, why is this the right thing to do? Are there other places in this file that have similar logic that it might be helpful to link to?